lib/stripe/util.rb

```diff
@@ -196,5 +197,44 @@ def self.check_api_key!(key)
       raise TypeError.new("api_key must be a string") unless key.is_a?(String)
       key
     end
+
+    private
+
+    # We use a pretty janky version of form encoding (Rack's) that supports
+    # more complex data structures like maps and arrays through the use of
+    # specialized syntax. To encode an array of maps like:
+    #
+    #     [{a: 1, b: 2}, {a: 3, b: 4}]
+    #
+    # We have to produce something that looks like this:
+    #
+    #     arr[][a]=1&arr[][b]=2&arr[][a]=3&arr[][b]=4
+    #
+    # The only way for the server to recognize that this is a two item array is
+    # that it notices the repetition of element "a", so it's key that these
+    # repeated elements are encoded first.
+    #
+    # This method is invoked for any arrays being encoded and checks that if
+    # the array contains all non-empty maps, that each of those maps must start
+    # with the same key so that their boundaries can be properly encoded.
+    def self.check_array_of_maps_start_keys!(arr)
+      expected_key = nil
+      arr.each do |item|
+        return if !item.is_a?(Hash)
+        return if item.count == 0
+
+        first_key = item.first[0]
```

Comment:
It's definitely a good idea to make this a recursive check, but luckily it already is! The check gets called from `.flatten_params` which will actually call into itself recursively if it detects another hash nested within it. This triggers the check again.

Worthwhile adding a test case to make sure that's true though. I did so in 7df3a4b.